# Adult Dataset

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
DATA_PATH = "/mnt/hdmenezess42/GitProjects/MLCourseAI/data/"
data = pd.read_csv(f"{DATA_PATH}adult.data.csv")

In [3]:
data.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country salary  
0          2174             0              40  United-States  <=50K  
1             0             0              13  United-States  <=50K  
2             0             0              40  United-States  <=50K  
3             0             0              40  United-States  <=50K  
4             0             0              40           Cuba  <=50K

In [4]:
print(data.keys())

Index(['age', 'workclass', 'fnlwgt', 'education', 'education-num',
       'marital-status', 'occupation', 'relationship', 'race', 'sex',
       'capital-gain', 'capital-loss', 'hours-per-week', 'native-country',
       'salary'],
      dtype='object')


## Data Shape

In [5]:
print(data.shape)

(32561, 15)


## Data Information

In [6]:
print(data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education-num   32561 non-null  int64 
 5   marital-status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital-gain    32561 non-null  int64 
 11  capital-loss    32561 non-null  int64 
 12  hours-per-week  32561 non-null  int64 
 13  native-country  32561 non-null  object
 14  salary          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB
None


## How many men and women are represented in this dataset?

In [7]:
data["sex"].describe(include=['object', 'bool'])

count     32561
unique        2
top        Male
freq      21790
Name: sex, dtype: object

In [8]:
data["sex"].value_counts()

sex
Male      21790
Female    10771
Name: count, dtype: int64

We can see in the cell above that we have 21790 man and 10771 woman

## What is the average age of women ?

- This involves two columns, sex=female and the age column.

In [9]:
data[data["sex"]=="Female"]["age"].mean()

np.float64(36.85823043357163)

The average age for woman is 37

## What is the percentage of German citizens ?

In [57]:
# My way
# data["native-country"].value_counts(normalize=True) # Wrong because show every one
float((data["native-country"] == "Germany").sum()) / data.shape[0]

0.004207487485028101

- Germany: 0.4%

## What are the mean and standard deviation of age for those who earn more than 50k per year and those who earn less than 50k per year?

In [11]:
data.groupby(["salary"])["age"].agg(["mean", "std"])

mean        std
salary                      
<=50K   36.783738  14.020088
>50K    44.249841  10.519028

## Is it true that people who earn more than 50k have at least high school education ?

- All unique values of education and it meaning:
    - Doctorate: doctoral degree (PhD, etc.)
    - Prof-school: professional degree (MD, JD, etc.)
    - Masters: master's degree
    - Bachelors: undergraduate degree
    - Assoc-acdm: academic associate degree
    - Assoc-voc: vocational associate degree
    - Some-college: attended college, no degree
    - HS-grad: high school graduate
    - 12th: 12th grade completed, no diploma
    - 11th: 11th grade completed
    - 10th: 10th grade completed
    - 9th: 9th grade completed
    - 7th-8th: completed 7th or 8th grade
    - 5th-6th: completed 5th or 6th grade
    - 1st-4th: completed 1st to 4th grade

    - Preschool: attended preschool only

In [12]:
data["education"].unique()

array(['Bachelors', 'HS-grad', '11th', 'Masters', '9th', 'Some-college',
       'Assoc-acdm', 'Assoc-voc', '7th-8th', 'Doctorate', 'Prof-school',
       '5th-6th', '10th', '1st-4th', 'Preschool', '12th'], dtype=object)

In [13]:
data["salary"].unique()

array(['<=50K', '>50K'], dtype=object)

In [14]:
data[data["salary"]==">50K"]["education"].value_counts(normalize=True)

education
Bachelors       0.283255
HS-grad         0.213621
Some-college    0.176891
Masters         0.122306
Prof-school     0.053947
Assoc-voc       0.046040
Doctorate       0.039026
Assoc-acdm      0.033797
10th            0.007907
11th            0.007652
7th-8th         0.005101
12th            0.004209
9th             0.003443
5th-6th         0.002041
1st-4th         0.000765
Name: proportion, dtype: float64

- Analyzing the data, we cannot confirm that higher education is directly linked to salaries above 50k. As we can see from the data, almost 30% are bachelor's degree holders, but we also have 21% of high school graduates. Meanwhile, the titles of professional school, associate vocational, doctorate, and associate academic are less than 5.5% each. 
- What we can conclude from this data is that a person has a greater chance of earning more than 50k if they have at least a high school degree, as all levels of education below this that earn 50k are less than 1% each.

## Display age statiscs for each race and each gender. Use groupby() and describe(). Find the maximum age of men of `Amer-Indian-Eskimo` race.

In [15]:
print(data['race'].unique())
print(data["race"].value_counts())

['White' 'Black' 'Asian-Pac-Islander' 'Amer-Indian-Eskimo' 'Other']
race
White                 27816
Black                  3124
Asian-Pac-Islander     1039
Amer-Indian-Eskimo      311
Other                   271
Name: count, dtype: int64


In [16]:
data['race'].describe(include=['object', 'bool'])

count     32561
unique        5
top       White
freq      27816
Name: race, dtype: object

In [17]:
print(data['sex'].unique())
print(data["sex"].value_counts())

['Male' 'Female']
sex
Male      21790
Female    10771
Name: count, dtype: int64


In [18]:
data.groupby(["race"])['age'].describe(percentiles=[])

count       mean        std   min   50%   max
race                                                               
Amer-Indian-Eskimo    311.0  37.173633  12.447130  17.0  35.0  82.0
Asian-Pac-Islander   1039.0  37.746872  12.825133  17.0  36.0  90.0
Black                3124.0  37.767926  12.759290  17.0  36.0  90.0
Other                 271.0  33.457565  11.538865  17.0  31.0  77.0
White               27816.0  38.769881  13.782306  17.0  37.0  90.0

In [19]:
data.groupby(["sex"])["age"].describe(percentiles=[])

count       mean        std   min   50%   max
sex                                                    
Female  10771.0  36.858230  14.013697  17.0  35.0  90.0
Male    21790.0  39.433547  13.370630  17.0  38.0  90.0

In [61]:
# I misinterpreted it and made statistical displays of age for each race and then for each sex. I didn't realize it was statistical displays of age for each race and sex, so here's the correct answer.
for (race, sex), sub_df in data.groupby(['race', 'sex']):
    print(f"Race: {0}, sex: {1}")
    print(sub_df['age'].describe())

Race: 0, sex: 1
count    119.000000
mean      37.117647
std       13.114991
min       17.000000
25%       27.000000
50%       36.000000
75%       46.000000
max       80.000000
Name: age, dtype: float64
Race: 0, sex: 1
count    192.000000
mean      37.208333
std       12.049563
min       17.000000
25%       28.000000
50%       35.000000
75%       45.000000
max       82.000000
Name: age, dtype: float64
Race: 0, sex: 1
count    346.000000
mean      35.089595
std       12.300845
min       17.000000
25%       25.000000
50%       33.000000
75%       43.750000
max       75.000000
Name: age, dtype: float64
Race: 0, sex: 1
count    693.000000
mean      39.073593
std       12.883944
min       18.000000
25%       29.000000
50%       37.000000
75%       46.000000
max       90.000000
Name: age, dtype: float64
Race: 0, sex: 1
count    1555.000000
mean       37.854019
std        12.637197
min        17.000000
25%        28.000000
50%        37.000000
75%        46.000000
max        90.000000
Name: ag

In [20]:
columns_to_show = ['race', 'sex']
data.groupby(['race', 'sex'])['age'].max()

race                sex   
Amer-Indian-Eskimo  Female    80
                    Male      82
Asian-Pac-Islander  Female    75
                    Male      90
Black               Female    90
                    Male      90
Other               Female    74
                    Male      77
White               Female    90
                    Male      90
Name: age, dtype: int64

## Among whom is the proportion of those who earn a lot greater: married or single men ? Consider as married those who have a marital-status starting with Married, the rest are considered bachelors.

- To solve this, it is easier if I create a new column `isMarried` with values 1 for married and 0 for single

In [21]:
data["isMarried"] = (
    data['marital-status'] .apply(lambda marital_status: 1 if 'Married-' in marital_status else 0)
)

In [22]:
data[["isMarried", 'marital-status']]

isMarried      marital-status
0              0       Never-married
1              1  Married-civ-spouse
2              0            Divorced
3              1  Married-civ-spouse
4              1  Married-civ-spouse
...          ...                 ...
32556          1  Married-civ-spouse
32557          1  Married-civ-spouse
32558          0             Widowed
32559          0       Never-married
32560          1  Married-civ-spouse

[32561 rows x 2 columns]

In [23]:
data[['salary', 'isMarried']].value_counts()
data.groupby(['isMarried'])[['sex', 'salary']].value_counts()

isMarried  sex     salary
0          Female  <=50K     8487
           Male    <=50K     7552
                   >50K       697
           Female  >50K       408
1          Male    <=50K     7576
                   >50K      5965
           Female  <=50K     1105
                   >50K       771
Name: count, dtype: int64

- We can conclude, observing the data above, that the number of married men who earn a lot (>50k) is much greater than single men with the same salary range.

## What is the maximum number of hours a person works per week (hours-per-week feature)? How many people work such a number of hours, and what is the percentage of those who earn a lot (>50k) among then ?

In [24]:
data.head()

age         workclass  fnlwgt  education  education-num  \
0   39         State-gov   77516  Bachelors             13   
1   50  Self-emp-not-inc   83311  Bachelors             13   
2   38           Private  215646    HS-grad              9   
3   53           Private  234721       11th              7   
4   28           Private  338409  Bachelors             13   

       marital-status         occupation   relationship   race     sex  \
0       Never-married       Adm-clerical  Not-in-family  White    Male   
1  Married-civ-spouse    Exec-managerial        Husband  White    Male   
2            Divorced  Handlers-cleaners  Not-in-family  White    Male   
3  Married-civ-spouse  Handlers-cleaners        Husband  Black    Male   
4  Married-civ-spouse     Prof-specialty           Wife  Black  Female   

   capital-gain  capital-loss  hours-per-week native-country salary  isMarried  
0          2174             0              40  United-States  <=50K          0  
1             0             0              13  United-States  <=50K          1  
2             0             0              40  United-States  <=50K          0  
3             0             0              40  United-States  <=50K          1  
4             0             0              40           Cuba  <=50K          1

In [ ]:
data['hours-per-week'].describe()

count    32561.000000
mean        40.437456
std         12.347429
min          1.000000
25%         40.000000
50%         40.000000
75%         45.000000
max         99.000000
Name: hours-per-week, dtype: float64

In [29]:
data['hours-per-week'].unique()

array([40, 13, 16, 45, 50, 80, 30, 35, 60, 20, 52, 44, 15, 25, 38, 43, 55,
       48, 58, 32, 70,  2, 22, 56, 41, 28, 36, 24, 46, 42, 12, 65,  1, 10,
       34, 75, 98, 33, 54,  8,  6, 64, 19, 18, 72,  5,  9, 47, 37, 21, 26,
       14,  4, 59,  7, 99, 53, 39, 62, 57, 78, 90, 66, 11, 49, 84,  3, 17,
       68, 27, 85, 31, 51, 77, 63, 23, 87, 88, 73, 89, 97, 94, 29, 96, 67,
       82, 86, 91, 81, 76, 92, 61, 74, 95])

In [30]:
data['hours-per-week'].max()

np.int64(99)

In [67]:
num_workhaloics = data[data['hours-per-week']==99].shape[0]
num_workhaloics

85

In [74]:
# data[data['hours-per-week']==99].groupby(['salary'])['hours-per-week'].value_counts()
rich_share = ( float( data[(data["hours-per-week"] == 99) & (data["salary"] == ">50K")].shape[0]) / num_workhaloics)
print(rich_share*100)

29.411764705882355


## Count the average time of work (hours-per-week) for those who earn a little and a lot (salary) for each contry (native-country). What will these be for japan ?

In [54]:
data.groupby(['salary'])['hours-per-week'].mean()

salary
<=50K    38.840210
>50K     45.473026
Name: hours-per-week, dtype: float64

In [51]:
data.groupby(['native-country','salary'])['hours-per-week'].mean()

native-country  salary
?               <=50K     40.164760
                >50K      45.547945
Cambodia        <=50K     41.416667
                >50K      40.000000
Canada          <=50K     37.914634
                            ...    
United-States   >50K      45.505369
Vietnam         <=50K     37.193548
                >50K      39.200000
Yugoslavia      <=50K     41.600000
                >50K      49.500000
Name: hours-per-week, Length: 82, dtype: float64

In [77]:
# Elegant method to do the same thing above
pd.crosstab(data['native-country'], data['salary'], values=data['hours-per-week'], aggfunc=np.mean).T

/tmp/nix-shell.b6So4o/ipykernel_8395/3999504109.py:2: FutureWarning: The provided callable <function mean at 0x7f3840008e00> is currently using DataFrameGroupBy.mean. In a future version of pandas, the provided callable will be used directly. To keep current behavior pass the string "mean" instead.
  pd.crosstab(data['native-country'], data['salary'], values=data['hours-per-week'], aggfunc=np.mean).T


native-country          ?   Cambodia     Canada      China   Columbia  \
salary                                                                  
<=50K           40.164760  41.416667  37.914634  37.381818  38.684211   
>50K            45.547945  40.000000  45.641026  38.900000  50.000000   

native-country       Cuba  Dominican-Republic    Ecuador  El-Salvador  \
salary                                                                  
<=50K           37.985714           42.338235  38.041667    36.030928   
>50K            42.440000           47.000000  48.750000    45.000000   

native-country    England  ...   Portugal  Puerto-Rico   Scotland     South  \
salary                     ...                                                
<=50K           40.483333  ...  41.939394    38.470588  39.444444  40.15625   
>50K            44.533333  ...  41.500000    39.416667  46.666667  51.43750   

native-country     Taiwan   Thailand  Trinadad&Tobago  United-States  \
salary                                                                 
<=50K           33.774194  42.866667        37.058824      38.799127   
>50K            46.800000  58.333333        40.000000      45.505369   

native-country    Vietnam  Yugoslavia  
salary                                 
<=50K           37.193548        41.6  
>50K            39.200000        49.5  

[2 rows x 42 columns]

In [78]:
data[data['native-country']=='Japan'].groupby(['native-country','salary'])['hours-per-week'].mean()

native-country  salary
Japan           <=50K     41.000000
                >50K      47.958333
Name: hours-per-week, dtype: float64